experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,662 @@
+    function InvariantRing(R::RepresentationReductiveGroup) #here G already contains information n and rep_mat
+        z = new()
+        if isdefined(R, :weights)
+            n = length(weights(R))
+            z.field = field(group(R))
+            super_ring, __ = graded_polynomial_ring(field(group(R)), "X"=>1:n)
+            z.poly_ring = super_ring
+            z.group = group(R)
+            z.representation = R
+            #z.fundamental = torus_invariants_fast(weights(R), ring)
+            return z
+        end
+        n = ncols(R.rep_mat)
+        z.representation = R
+        z.group = R.group
+        G = z.group
+        z.field = G.field
+        z.poly_ring, __ = graded_polynomial_ring(G.field, "X" => 1:n)
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+    
+    #to compute invariant ring ring^G where G is the reductive group of R. 
+    function InvariantRing(R::RepresentationReductiveGroup, ring::MPolyDecRing)
+        n = ncols(R.rep_mat)
+        n == ngens(ring) || error("The given polynomial ring is not compatible.")
+        z = new()
+        if isdefined(R, :weights)
+            #dosomething
+        end
+        z.representation = R
+        z.group = R.group
+        G = R.group
+        z.field = G.field
+        z.poly_ring = ring
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+end
+
+invariant_ring(R::RepresentationReductiveGroup) = InvariantRing(R)
+invariant_ring(ring::MPolyDecRing, R::RepresentationReductiveGroup) = InvariantRing(R, ring)
+null_cone_ideal(R::InvariantRing) = R.NullConeIdeal
+poly_ring(R::InvariantRing) = R.poly_ring
+
+function fundamental_invariants(z::InvariantRing)
+    if isdefined(z, :fundamental)
+        return z.fundamental
+    else
+        R = z.representation
+        if group(group(R))[1] == :torus && isdefined(R, :weights)
+            z.fundamental = torus_invariants_fast(weights(R), poly_ring(z))
+            return z.fundamental
+        end
+        I, M = proj_of_image_ideal(R.group, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(R.group, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, R.group, z.poly_ring, M, I, z.reynolds_operator)
+        return z.fundamental
+    end
+end
+
+function Base.show(io::IO, R::InvariantRing) 
+    io = AbstractAlgebra.pretty(io)
+    println(io, "Invariant Ring of")
+    show(io, R.poly_ring)
+    print(io, AbstractAlgebra.Indent(),  " under group action of ", R.group.group[1], R.group.group[2])
+    print(io, AbstractAlgebra.Dedent())
+end
+
+#computing the graph Gamma from Derksens paper
+function image_ideal(G::ReductiveGroup, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
+    R = base_ring(rep_mat)
+    n = ncols(rep_mat)
+    m = G.group[2]
+    mixed_ring_xy, x, y, zz = polynomial_ring(G.field, "x"=>1:n, "y"=>1:n, "zz"=>1:m^2)
+    ztozz = hom(R,mixed_ring_xy, gens(mixed_ring_xy)[(2*n)+1:(2*n)+(m^2)])
```

Comment:
If you reorder the variables, the map gets a bit simpler
```suggestion
    mixed_ring_xy, zz, x, y = polynomial_ring(G.field, "zz"=>1:m^2, "x"=>1:n, "y"=>1:n)
    ztozz = hom(R,mixed_ring_xy, gens(mixed_ring_xy)[1:m^2])
```